### PR TITLE
Get evil initial state from parent modes [updated]

### PR DESCRIPTION
**Pull request by VanLaser**
This is basically a modification of the solution proposed by @wasamasa (described in #775 and also discussed in #497) which allows `evil-initial-state` to decide the initial state for buffers based not only on the state set for the current mode (e.g. using `evil-set-initial-state`), but also on those of its parent modes. 

I added a check for unusual situations in the parent -> derived modes relationship (see [Malabarba/paradox#86](https://github.com/Malabarba/paradox/issues/86)), in which a derived mode is set also as a parent of its parent; in such situations, `evil-initial-state` will now raise an error (avoiding the endless loop from the original solution). Another proposed [approach](https://github.com/emacs-evil/evil/issues/497#issuecomment-322514176) was that of using a counter that bails out with an error if the visit from a parent to its parent (and so on) goes above a limit, but I think the current solution is the best because it checks for the actual problem.

I also added an extra optional parameter (`noinherit`) for `evil-initial-state`, allowing for an initial state check that does not consider the parents of the mode, just the mode itself (i.e. in this case, `evil-initial-state` works exactly as before this patch).

### CODE REVIEW

File: evil-core.el [hunk 1]
```diff
@@ -285,17 +285,39 @@ See also `evil-initial-state'."
...
-The initial state for a mode can be set with
+Returns DEFAULT if no initial state is associated with either
+MODE, or any of its parents. If NOINHERIT is not nil, returns
+DEFAULT if no initial state is associated with this mode. The
+initial state for a mode can be set with
 `evil-set-initial-state'."
   (let (state modes)
     (catch 'done
-      (dolist (entry (evil-state-property t :modes) default)
+      (dolist (entry (nreverse (evil-state-property t :modes)) default)
```
Comment (wasamasa) [inline evil-core.el]: I'm not sure why I did add a `nreverse` here, IIRC it was so that my default state doesn't kick in yet.

Comment (VanLaser) [inline evil-core.el]: It's true that I didn't follow that to the end, I just assumed you had a good reason for it. The test passes (on my PC) even without it. Should I take it out?

Comment (wasamasa) [inline evil-core.el]: Yeah, if the need still arises, it would be better to make the order modes are checked customizable.

**Comment by VanLaser** [comment]
Ok, please don't pull it yet; I thought about that order and I think I'll try to rework the logic, so that the code will first try to get the list of the parents (once) and then it will check upwards for the initial state of each parent, until I find one (or not). In any case, the initial state would come from the closest parent in the hierarchy (e.g. "father", not "grandfather"), rather than from the order given by the `(evil-state-property :modes)`. Since a mode can't be set to two initial states at the same time (I suppose!), that order shouldn't count in this case. I'll probably make a new clear pull request (I'm not too used with this github workflow).

**Comment by VanLaser** [comment]
Should be OK now.

**Pull request closed without merging**